modules/tinymce/src/core/main/ts/selection/MultiClickSelection.ts

```diff
@@ -1,44 +1,82 @@
 import { SugarElement } from '@ephox/sugar';
 
+import DomTreeWalker from '../api/dom/TreeWalker';
 import Editor from '../api/Editor';
-import * as CaretFinder from '../caret/CaretFinder';
+import { isCaretCandidate } from '../caret/CaretCandidate';
 import { CaretPosition } from '../caret/CaretPosition';
-import { isTextBlock } from '../dom/ElementType';
+import { isBlock } from '../dom/ElementType';
 import * as NodeType from '../dom/NodeType';
 import * as RangeNormalizer from './RangeNormalizer';
 
-const isTextBlockNode = (node: Node): node is Element => NodeType.isElement(node) && isTextBlock(SugarElement.fromDom(node));
+const isBr = NodeType.isBr;
+const isText = NodeType.isText;
 
-const normalizeSelection = (editor: Editor) => {
-  const rng = editor.selection.getRng();
-  const startPos = CaretPosition.fromRangeStart(rng);
-  const endPos = CaretPosition.fromRangeEnd(rng);
+const isBoundary = (node: Node): boolean => {
+  return isBr(node) || isBlock(SugarElement.fromDom(node));
+};
 
-  if (CaretPosition.isElementPosition(startPos)) {
-    const container = startPos.container();
-    if (isTextBlockNode(container)) {
-      CaretFinder.firstPositionIn(container).each((pos) => rng.setStart(pos.container(), pos.offset()));
+const walkBackwardWhile = (startNode: Node, rootNode: Node): Node => {
```

Comment:
I wonder if these two could be the same function but maybe that will just make the code more confusing. Another thing is that it should probably also not select the last caret candidate if it finds one within a noneditable element. So for this case `foo<span contenteditable="false">bar</span>` the selection should be a element position after the noneditable span not the text node within it. I think this could just be a separate step where you check if the found nodes closest contentEditable state is false then place the point after that element. I'm guessing some browsers will normalize the selection for us in this case anyway moving it outside.